### ISSUE TITLE
How to set up a training set in a supervised method?

### ISSUE THREAD

**Issue by qyangcauc** [boudinfl/pke#20]
In readme, I know how to set up the training file, for example，I have n papers for training file.But I don't know  how to use the key words that the author of this n paper has tagged.In examples/train-model.py，I understand the following code.
      pke.train_supervised_model(input_dir=input_dir,
                           reference_file=reference_file,
                           model_file=output_mdl,
                           df=df_counts,
                           format="corenlp",
                           use_lemmas=False,
                           stemmer="porter",
                           model=pke.supervised.Kea(),
                           language='english',
                           extension="xml")
input_dir represents a collection of n papers.I don’t know what reference_file and use_lemmas means, and how to used the author’s keyphrases of n papers in a supervised method to training model of kea.Because   
You only specify n papers , without the keywords of these papers.I don't know if reference_file is the keyword document.So，I am eager to know how the keywords are used in the code, I wish you would tell me in detail that I would be grateful.

**Comment by boudinfl** [comment]
Hi @qyangcauc 

The input should contain the documents, each document in XML format (already pre-processed using Stanford CoreNLP).

The reference file contain the keywords for each document in the SemEval-2010 format. Details about this format are available at http://docs.google.com/Doc?id=ddshp584_46gqkkjng4

You can look at the SemEval data for an example: http://semeval2.fbk.eu/semeval2.php?location=data

the `use_lemmas` parameter allows to use the Lemmas produced by Stanford CoreNLP instead of stems generated with `nltk`. For most uses, it is better to let it to False.

Best regards,

f.

**Comment by qyangcauc** [comment]
Dear boudinfl

The problem：
I already know how to train with your help, but you only write how to get keywords with unsupervised in readme. I already trian model with my set, when I visit visit https://boudinfl.github.io/pke/build/html/supervised.html, I don't know how to use the model I've trained.

    This is my copy in unsupervised for my model:

    # test file, wait for the keywords to be extracted.
    extractor = pke.supervised.Kea(input_file='./dataset/4046.txt')
    # test set format
    extractor.read_document(format='preprocessed')
    # test set df count
   df_counts = pke.load_document_frequency_file(input_file='./dataset/df.txt')  #
   extractor.feature_extraction(df=df_counts)
   # kea model that have been trained
   extractor.classify_candidates(model='./dataset/kea_model/kea_kdd_model')     .
   # I don't know how to get keywords
   print extractor.get_n_best(n=2)

I know it's wrong. So I want to you tell me how can I get keywords in test set in supervised models with that have been trained.

Debatable：
After I calculate df, and then write to the file df_count.txt, I found the wrong encoding format.That is a mess. In function compute_document_frequency() , you use 
    
   with gzip.open(output_file, 'w') as f:
        f.write('--NB_DOC--' + delimiter + str(nb_documents) + '\n')
        for ngram in frequencies:
            f.write((ngram).encode('utf-8') + delimiter + str(len(frequencies[ngram])) + '\n')

when I use your code,  I found the wrong encoding format.That is a mess in my df_count.txt.I tried to modify the encoding, but it also happened. So I modified that：
    with open(output_file, 'w') as f:
This is successfully . My running environment：win10 & Pycharm & Python2.7 . I don't know why that is happening.

Best regards,
qyang.

**Comment by qyangcauc** [comment]
Dear boudinfl

The problem：
I already know how to train with your help, but you only write how to get keywords with unsupervised in readme. I already trian model with my set, when I visit visit https://boudinfl.github.io/pke/build/html/supervised.html, I don't know how to use the model I've trained.

This is my copy in unsupervised for my model:

# test file, wait for the keywords to be extracted.
extractor = pke.supervised.Kea(input_file='./dataset/4046.txt')
# test set format
extractor.read_document(format='preprocessed')
# test set df count
df_counts = pke.load_document_frequency_file(input_file='./dataset/df.txt') #
extractor.feature_extraction(df=df_counts)
#kea model that have been trained
extractor.classify_candidates(model='./dataset/kea_model/kea_kdd_model') .
# I don't know how to get keywords
print extractor.get_n_best(n=2)

I know it's wrong. So I want to you tell me how can I get keywords in test set in supervised models with that have been trained.

Debatable：
After I calculate df, and then write to the file df_count.txt, I found the wrong encoding format.That is a mess. In function compute_document_frequency() , you use

with gzip.open(output_file, 'w') as f:
f.write('--NB_DOC--' + delimiter + str(nb_documents) + '\n')
for ngram in frequencies:
f.write((ngram).encode('utf-8') + delimiter + str(len(frequencies[ngram])) + '\n')

when I use your code, I found the wrong encoding format.That is a mess in my df_count.txt.I tried to modify the encoding, but it also happened. So I modified that：
with open(output_file, 'w') as f:
This is successfully . My running environment：win10 & Pycharm & Python2.7 . I don't know why that is happening.

Best regards,
qyang.

**Comment by qyangcauc** [comment]
Dear boudinfl

The problem：
I already know how to train with your help, but you only write how to get keywords with unsupervised in readme. I already trian model with my set, when I visit visit https://boudinfl.github.io/pke/build/html/supervised.html, I don't know how to use the model that I've trained.

This is my copy in unsupervised for my model:
```
#test file, wait for the keywords to be extracted.
extractor = pke.supervised.Kea(input_file='./dataset/4046.txt')
#test` set format
extractor.read_document(format='preprocessed')
#test set df count
df_counts = pke.load_document_frequency_file(input_file='./dataset/df.txt') #
extractor.feature_extraction(df=df_counts)
#kea model that have been trained
extractor.classify_candidates(model='./dataset/kea_model/kea_kdd_model') .
#I don't know how to get keywords
print extractor.get_n_best(n=2)
```
I know it's wrong. So I want to you tell me how can I get keywords in test set in supervised models with that I have been trained.

Debatable：
After I calculate df, and then write to the file df_count.txt, I found the wrong encoding format.That is a mess. In function compute_document_frequency() , you use

```
with gzip.open(output_file, 'w') as f:
    f.write('--NB_DOC--' + delimiter + str(nb_documents) + '\n')
    for ngram in frequencies:
        f.write((ngram).encode('utf-8') + delimiter + str(len(frequencies[ngram])) + '\n')
```

when I use your code, I found the wrong encoding format.That is a mess in my df_count.txt.I tried to modify the encoding, but it also happened. So I modified that：

`with open(output_file, 'w') as f:
    ...`

This is successfully . My running environment：win10 & Pycharm & Python2.7 . I don't know why that is happening.

Best regards,
qyang.

**Comment by boudinfl** [comment]
So for supervised models (this example is for `Kea`), the API is as follows:

``
import pke

extractor = pke.supervised.Kea(input_file='/path/to/input')

# here the input document should be in preprocessed format, i.e. 
# whitespace-separated POS-tagged tokens, one sentence per line.
extractor.read_document(format='preprocessed')

# load the df counts
# here the Df counts file is normally a gzip compressed file, this is why you have the issue in the
# second part of your comment
df_counts = pke.load_document_frequency_file(input_file='/path/to/dfcounts')

# Extract the candidates, I think you miss that
extractor.candidate_selection()

# Compute the features for the candidates
extractor.feature_extraction(df=df_counts)

# Classify the candidates using your trained model
extractor.classify_candidates(model='./dataset/kea_model/kea_kdd_model') .

Get the best 2 candidates
print extractor.get_n_best(n=2)
``
Best,

f.
